### PR TITLE
add clip_grad_norm_ API

### PR DESCRIPTION
**Pull request by zxcd**
### PR types

New features 

### PR changes

APIs

### Describe

add clip_grad_norm API and unit test.

this API can only run in dynamic mode, cause only dynamic graph supports gradient, and 'Variable' Object not supported .grad

### CODE REVIEW

File: python/paddle/nn/utils/clip_grad_norm_.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+
+__all__ = []
```
Comment (zh794390558) [inline python/paddle/nn/utils/clip_grad_norm_.py]: clip_grad_norm_ 加进列表里。

等CI都过后再push新的commit。

File: python/paddle/nn/utils/clip_grad_norm_.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+
+def clip_grad_norm_(
```
Comment (zh794390558) [inline python/paddle/nn/utils/clip_grad_norm_.py:20]: 加上no_grad的guard？

File: python/paddle/fluid/tests/unittests/test_clip_grad_norm_.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            input_pd = paddle.to_tensor(
+                np.random.random([1, 2]).astype(np.float32)
+            )
+            input_pd.grad = paddle.to_tensor(
+                np.random.random([1, 2]).astype(np.float32)
+            )
+            clip_grad_norm_(input_pd, max_norm=2, norm_type=float("inf"))
+
+        self.assertRaises(RuntimeError, TestRuntimeErrorStaticMode)
+        paddle.disable_static()
```
Comment (zh794390558) [inline python/paddle/fluid/tests/unittests/test_clip_grad_norm_.py]: 放到TestRuntimeErrorStaticMode里？

File: python/paddle/nn/utils/clip_grad_norm_.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    max_norm: float,
+    norm_type: float = 2.0,
+    error_if_nonfinite: bool = False,
+) -> paddle.Tensor:
+    r"""Clips gradient norm of an iterable of parameters.
+
+    The norm is computed over all gradients together, as if they were
+    concatenated into a single vector. Gradients are modified in-place.
+
+    This API can only run in dynamic mode, not static mode.
```
Comment (dingjiaweiww) [inline python/paddle/nn/utils/clip_grad_norm_.py]: 改成：dynamic graph mode

File: python/paddle/nn/utils/clip_grad_norm_.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        Total norm of the parameter gradients (viewed as a single vector).
+    Example:
+        .. code-block:: python
+            import paddle
+
+            x = paddle.uniform([10, 10], min=-1.0, max=1.0, dtype='float32')
+            max_norm = float(5.0)
+            linear = paddle.nn.Linear(in_features=10, out_features=10,
+                                      weight_attr=paddle.ParamAttr(need_clip=True),
+                                      bias_attr=paddle.ParamAttr(need_clip=False))
```
Comment (sneaxiy) [inline python/paddle/nn/utils/clip_grad_norm_.py]: Why set `need_clip` here?

File: python/paddle/nn/utils/clip_grad_norm_.py (hunk 4 above)
Comment (zxcd) [reply python/paddle/nn/utils/clip_grad_norm_.py]: 好的

File: python/paddle/nn/utils/clip_grad_norm_.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            import paddle
+
+            x = paddle.uniform([10, 10], min=-1.0, max=1.0, dtype='float32')
+            max_norm = float(5.0)
+            linear = paddle.nn.Linear(in_features=10, out_features=10)
+            out = linear(x)
+            loss = paddle.mean(out)
+            loss.backward()
+
+            clip_grad_norm_(linear.parameters(), max_norm)
```
Comment (jzhang533) [inline python/paddle/nn/utils/clip_grad_norm_.py]: 这行写API的全路径吧

File: python/paddle/nn/utils/clip_grad_norm_.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+__all__ = ['clip_grad_norm_']
+
+
+@paddle.no_grad()
+def clip_grad_norm_(
+    parameters: paddle.Tensor,
+    max_norm: float,
+    norm_type: float = 2.0,
+    error_if_nonfinite: bool = False,
+) -> paddle.Tensor:
```
Comment (jzhang533) [inline python/paddle/nn/utils/clip_grad_norm_.py]: 这个type information加的不对吧。
parameters 参数应该是 list of tensors ？
paddle的API，99%都没有加type information，我觉得这个API也可以先不加。

**Review by jzhang533 — APPROVED** [review]
LGTM

**Review by dingjiaweiww — APPROVED** [review]
LGTM

**Review by sneaxiy — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]
LGTM for API defination
